### PR TITLE
Add beanstalk driver

### PR DESCRIPTION
**Pull request by ameech**
Adds Beanstalk support using https://github.com/pda/pheanstalk

Also, added a method to mark jobs as processed to provide more reliable queues. The Redis driver will need to be updated to be more reliable in case of failures, such as network and memory related.

### CODE REVIEW

File: src/Driver/RedisDriver.php [hunk 1]
```diff
@@ -34,6 +34,17 @@ public function dequeue($queue)
...
+            null,
+        ];
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function processed($job)
+    {
+        return true;
```
Comment (ameech) [inline src/Driver/RedisDriver.php:48]: Need to look into supporting: http://redis.io/commands/rpoplpush#pattern-reliable-queue

File: src/Worker.php [hunk 2]
```diff
@@ -84,18 +84,18 @@ protected function tick($queue)
...
+        list($command, $job) = $this->driver->dequeue($queue);
+        if (empty($command) && $this->drain) {
             $this->event->drained();
             return false;
-        } elseif (empty($message)) {
+        } elseif (empty($command)) {
             return true;
         }
 
-        $command = unserialize($message);
```
Comment (ameech) [inline src/Worker.php:95]: This is the responsibility of the driver, as the serialization already happens there.

File: src/Driver/BeanstalkDriver.php [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+    /**
+     * @inheritdoc
+     */
+    public function dequeue($queue)
+    {
+        $job = $this->beanstalk
+            ->watch($queue)
+            ->ignore('default')
+            ->reserve(5);
```
Comment (fentie) [inline src/Driver/BeanstalkDriver.php]: what makes 5 special?

Comment (ameech) [reply src/Driver/BeanstalkDriver.php]: It's just the time out, it's what I've used within the Redis driver.

Comment (fentie) [reply src/Driver/BeanstalkDriver.php]: make it a constant?

Comment (ameech) [reply src/Driver/BeanstalkDriver.php]: Certainly could.

Comment (ameech) [reply src/Driver/BeanstalkDriver.php]: @fentie added.

File: src/Driver/RedisDriver.php [hunk 4]
```diff
@@ -32,8 +32,19 @@ public function enqueue($queue, $command)
      */
     public function dequeue($queue)
     {
-        list($_, $message) = $this->redis->blPop($queue, 5) ?: null;
+        list($_, $message) = $this->redis->blPop($queue, static::TIMEOUT) ?: null;
```
Comment (elazar) [inline src/Driver/RedisDriver.php]: `$_` isn't needed. This is perfectly legal:

``` php
list(, $message) = $this->redis->blPop($queue, static::TIMEOUT) ?: null;
```

File: tests/Driver/BeanstalkDriverTest.php [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+
+        $this->assertEquals([
+            false,
+            false,
+        ], $this->driver->dequeue('test-queue'));
+    }
+
+    public function testProcessed()
+    {
+        $job = new \stdClass();
```
Comment (elazar) [inline tests/Driver/BeanstalkDriverTest.php]: Nitpick: parentheses shouldn't be necessary if you aren't passing arguments. I'd stay with consistent with this (see the first line of the previous test).

**Review by elazar — CHANGES_REQUESTED** [review]
Overall, looks fine. Just a few nitpicks.

File: src/Driver/RedisDriver.php (hunk 4 above)
Comment (ameech) [reply src/Driver/RedisDriver.php]: Both ways are ugly and I wish there was a better way to do it, but I can drop the `$_` as it just pollutes.

File: tests/Driver/BeanstalkDriverTest.php (hunk 5 above)
Comment (ameech) [reply tests/Driver/BeanstalkDriverTest.php]: Whoops, missed that one.

**Pull request merged**